File: Common/Sound/src/common/util/PlaySound.java
/**
 * @author Julian Sender
 */
package common.util;

import javafx.application.Platform;
import javafx.scene.media.Media;
import javafx.scene.media.MediaException;
import javafx.scene.media.MediaPlayer;
import javafx.util.Duration;

/**
 * Class to play sounds
 */
public final class PlaySound {

    private static final Player p = new Player();

    /**
     *
     * @param path Player needs the path of mediafile as parameter
     */
    public static void playSound( final String path ) {
        p.playSound( path, false );
    }

    /**
     * Plays and resets sound
     * @param path Player needs the path of mediafile as parameter
     */
    public static void playAndResetSound( final String path ) {
        p.playSound( path, true );
    }

    /**
     * Plays a sound
     * @param path Player needs the path of mediafile as parameter
     * @param resetMediaPlayer true, if it's desired to reset mediaplayer
     */
    public static void playSound( final String path, final boolean resetMediaPlayer ) {
        p.playSound( path, resetMediaPlayer );
    }

    /**
     * Check if next sound is different to previous, else save performance :-)
     */
    static class Player implements Runnable {
        private String      musicFile    = null;
        private Media       currentMedia = null;
        private MediaPlayer prop_mediaPlayer;
        private boolean     resetTimer;

        public void playSound( final String path, final boolean resetMediaPlayer ) throws MediaException{
            initProperties(path, resetMediaPlayer);
            if (prop_mediaPlayer != null) {
                if ( resetTimer ) {
                    resetTimer( prop_mediaPlayer );
                }
                Platform.runLater( () -> prop_mediaPlayer.play() );
            }
            // else { throw new NullPointerException("Mediaplayer is null"); }
        }

        /**
         * Resets timer for MediaPlayer
         * @param mediaPlayer needs instance of mediaplayer where to reset timer
         */
        private void resetTimer( MediaPlayer mediaPlayer ) {
            if ( mediaPlayer.getCurrentTime().lessThan( mediaPlayer.getTotalDuration() ) ) {
                mediaPlayer.stop();
                mediaPlayer.seek( Duration.ZERO );
            }
        }

        private void initProperties(final String path, final boolean resetMediaPlayer ) {
            resetTimer = resetMediaPlayer;
            if (musicFile != null && musicFile.equals(path)) { //Performance-Kniff
                return;
            }
            if ( prop_mediaPlayer != null ) {
                prop_mediaPlayer.dispose();
            }
            musicFile = path;
            if (musicFile == null) {
                throw new NullPointerException("MusicFile String is null");
            }
            currentMedia = AudioBuffer.loadMedia( path );
            if (currentMedia == null) {
                throw new NullPointerException("Current Media is null: " + musicFile);
            }
            prop_mediaPlayer = createMediaPlayer();
        }

        /**
         * Creates isntance of Mediaplayer
         * @return returns the new instance of mediaplayer
         */
        private MediaPlayer createMediaPlayer() {
            MediaPlayer mediaPlayer = new MediaPlayer( currentMedia );
            mediaPlayer.setAutoPlay( false );
            mediaPlayer.setOnEndOfMedia( () -> {
                mediaPlayer.stop();
                mediaPlayer.seek( Duration.ZERO );
                //mediaPlayer.dispose();
            } );
            return mediaPlayer;
        }

        /**
         * Lets it all work together
         */
        @Override
        public void run() {
            playSound( musicFile, resetTimer );
        }
    }



}

File: Common/Sound/src/common/util/PlaySound.kt
/**
 * @author Julian Sender
 */
package common.util

import javafx.application.Platform
import javafx.scene.media.Media
import javafx.scene.media.MediaException
import javafx.scene.media.MediaPlayer
import javafx.util.Duration

/**
 * Class to play sounds
 */
object PlaySound {
    private val p = Player()

    /**
     * Buffer Media before it needs to be played.
     */
    fun loadMedia(mediaPath: String): Media = AudioBuffer.loadMedia(mediaPath)!!

    /**
     * Plays a sound
     * @param path Player needs the path of mediafile as parameter
     * @param resetMediaPlayer true, if it's desired to reset mediaplayer
     */
    fun playSound(path: String, resetMediaPlayer: Boolean = false) = p.playSound(path, resetMediaPlayer)

    /**
     * Plays and resets sound
     * @param path Player needs the path of mediafile as parameter
     */
    fun playAndResetSound(path: String) = p.playSound(path, true)

    fun stop() = p.stop()

    fun pause() = this.p.pause()

    fun resume() = this.p.resume()

    /**
     * Check if next sound is different to previous, else save performance :-)
     */
    internal class Player : Runnable {

        private var musicFile: String? = null
        private var currentMedia: Media? = null
        private var propMediaplayer: MediaPlayer? = null
        private var resetTimer = false

        fun stop() = this.propMediaplayer?.stop()
        fun pause() = this.propMediaplayer?.pause()
        fun resume() = this.propMediaplayer?.play()

        @Throws(MediaException::class)
        fun playSound(path: String, resetMediaPlayer: Boolean) {

            fun createMediaPlayer(): MediaPlayer {
                val mediaPlayer = MediaPlayer(currentMedia)
                mediaPlayer.isAutoPlay = false
                mediaPlayer.onEndOfMedia = Runnable {
                    mediaPlayer.stop()
                    mediaPlayer.seek(Duration.ZERO)
                }
                return mediaPlayer
            }

            resetTimer = resetMediaPlayer
            if (musicFile != null && musicFile == path) { //Performance-Kniff
                return
            }
            if (propMediaplayer != null) propMediaplayer!!.dispose()
            musicFile = path
            currentMedia = AudioBuffer.loadMedia(path)
            propMediaplayer = createMediaPlayer()

            if (propMediaplayer != null) {
                if (resetTimer) {
                    resetTimer(propMediaplayer!!)
                }
                Platform.runLater { propMediaplayer!!.play() }
            }
        }

        /**
         * Resets timer for MediaPlayer
         * @param mediaPlayer needs instance of mediaplayer where to reset timer
         */
        private fun resetTimer(mediaPlayer: MediaPlayer) {
            if (mediaPlayer.currentTime.lessThan(mediaPlayer.totalDuration)) {
                mediaPlayer.stop()
                mediaPlayer.seek(Duration.ZERO)
            }
        }

        /**
         * Lets it all work together
         */
        override fun run() {
            musicFile?.let { playSound(it, resetTimer) }
        }
    }
}
